src/backend/pipeline/cont_scheduler.c

```diff
@@ -703,12 +720,18 @@ ContQuerySchedulerMain(int argc, char *argv[])
 		 * Wait until naptime expires or we get some type of signal (all the
 		 * signal handlers will wake us by calling SetLatch).
 		 */
-		rc = WaitLatch(&MyProc->procLatch, WL_LATCH_SET | WL_POSTMASTER_DEATH, 0);
+		rc = WaitLatch(&MyProc->procLatch, WL_LATCH_SET | WL_POSTMASTER_DEATH | WL_TIMEOUT, 1000);
 
 		ResetLatch(&MyProc->procLatch);
 
 		DisableCatchupInterrupt();
 
+		if (rc & WL_TIMEOUT)
```

Comment:
AdhocMgrPeriodicDrain is probably a better name. It is currently possible to insert tuples into the adhoc tuple buffer that won't be read (because the adhoc process has finished). The Drain function scans the adhoc tuple buffer for these kinds of tuples and removes them.

If it is not called, eventually the buffer will fill up and deadlock.

A better design or better synchronization between reader / writers could obviate this.